File: apps/web/src/engine/heapSort.ts
import type { AlgorithmSnapshot } from '@dsa-tutor/types'
import { CriticalJunctionType, JunctionDifficulty, PredictionType } from '@dsa-tutor/types'

export interface HeapSortState {
  array: number[]
  /** Elements at index >= heapSize are already in their final sorted position. */
  heapSize: number
}

const PSEUDOCODE_LINE = {
  BUILD_START: 0,
  HEAPIFY_DEF: 1,
  INIT_LARGEST: 2,
  COMPARE: 3,
  SWAP: 4,
  EXTRACT_LOOP: 5,
  EXTRACT_SWAP: 6,
  SIFT_DOWN: 7,
  DONE: 8,
} as const

interface SnapshotParams {
  stepIndex: number
  description: string
  pseudocodeLine: number
  isPredictionRequired: boolean
  state: HeapSortState
  activeIndices?: number[]
  highlightIndices?: number[]
  comparedIndices?: number[]
  swappedIndices?: number[]
  isFinalStep?: boolean
  criticalJunctionType?: CriticalJunctionType | null
  junctionDifficulty?: JunctionDifficulty | null
}

function makeSnapshot(params: SnapshotParams): AlgorithmSnapshot {
  return {
    stepIndex: params.stepIndex,
    description: params.description,
    pseudocodeLine: params.pseudocodeLine,
    isPredictionRequired: params.isPredictionRequired,
    predictionType: PredictionType.TILE_GRID,
    dataStructureState: { ...params.state, array: [...params.state.array] },
    activeIndices: [...(params.activeIndices ?? [])],
    highlightIndices: [...(params.highlightIndices ?? [])],
    comparedIndices: [...(params.comparedIndices ?? [])],
    swappedIndices: [...(params.swappedIndices ?? [])],
    isFinalStep: params.isFinalStep ?? false,
    criticalJunctionType: params.criticalJunctionType ?? null,
    junctionDifficulty: params.junctionDifficulty ?? null,
  }
}

/**
 * Pure snapshot engine for Heap Sort. Two phases: build a max-heap by
 * heapifying every internal node bottom-up, then repeatedly swap the
 * root (the current maximum) with the last unsorted element and sift
 * the new root down to restore the heap property. Never mutates
 * `input`.
 */
export function heapSortEngine(input: number[]): AlgorithmSnapshot[] {
  const working = [...input]
  const n = working.length
  const snapshots: AlgorithmSnapshot[] = []
  let stepIndex = 0

  function push(params: Omit<SnapshotParams, 'stepIndex'>) {
    snapshots.push(makeSnapshot({ ...params, stepIndex: stepIndex++ }))
  }

  // Sifts the element at index `i` down within the heap bounded by
  // `size`, comparing it against its larger child at every level where
  // a comparison is actually meaningful (a leaf has nothing to compare
  // against, so it's skipped silently rather than padding the step
  // sequence with a no-op snapshot).
  function heapify(i: number, size: number) {
    const left = 2 * i + 1
    const right = 2 * i + 2

    let largerChild: number
    if (left < size && right < size) {
      largerChild = working[left] >= working[right] ? left : right
    } else if (left < size) {
      largerChild = left
    } else if (right < size) {
      largerChild = right
    } else {
      return
    }

    const shouldSift = working[largerChild] > working[i]

    push({
      description: `Heapify: parent index ${i} (value ${working[i]}) vs child index ${largerChild} (value ${working[largerChild]}). Should the parent sift down?`,
      pseudocodeLine: PSEUDOCODE_LINE.COMPARE,
      isPredictionRequired: true,
      state: { array: working, heapSize: size },
      activeIndices: [i, largerChild],
      comparedIndices: [largerChild],
      criticalJunctionType: CriticalJunctionType.HEAP_COMPARE,
      junctionDifficulty: JunctionDifficulty.PROCEDURAL,
    })

    if (!shouldSift) {
      push({
        description: `${working[i]} is already at least as large as both children. The heap property holds here.`,
        pseudocodeLine: PSEUDOCODE_LINE.INIT_LARGEST,
        isPredictionRequired: false,
        state: { array: working, heapSize: size },
        activeIndices: [i],
      })
      return
    }

    const temp = working[i]
    working[i] = working[largerChild]
    working[largerChild] = temp

    push({
      description: `Swapped index ${i} and index ${largerChild}. The array is now [${working.join(', ')}].`,
      pseudocodeLine: PSEUDOCODE_LINE.SWAP,
      isPredictionRequired: false,
      state: { array: working, heapSize: size },
      swappedIndices: [i, largerChild],
    })

    heapify(largerChild, size)
  }

  push({
    description:
      'Starting Heap Sort. First, the array is rearranged into a max-heap: every parent is at least as large as its children.',
    pseudocodeLine: PSEUDOCODE_LINE.BUILD_START,
    isPredictionRequired: false,
    state: { array: working, heapSize: n },
  })

  for (let i = Math.floor(n / 2) - 1; i >= 0; i--) {
    push({
      description: `Heapifying the subtree rooted at index ${i}.`,
      pseudocodeLine: PSEUDOCODE_LINE.HEAPIFY_DEF,
      isPredictionRequired: false,
      state: { array: working, heapSize: n },
      activeIndices: [i],
    })
    heapify(i, n)
  }

  push({
    description: `The max-heap is built: [${working.join(', ')}]. The root, index 0, now holds the maximum value.`,
    pseudocodeLine: PSEUDOCODE_LINE.EXTRACT_LOOP,
    isPredictionRequired: false,
    state: { array: working, heapSize: n },
  })

  for (let i = n - 1; i >= 1; i--) {
    push({
      description: `Extract ${working[0]} — it's the current maximum.`,
      pseudocodeLine: PSEUDOCODE_LINE.EXTRACT_SWAP,
      isPredictionRequired: true,
      state: { array: working, heapSize: i + 1 },
      activeIndices: [0],
      criticalJunctionType: CriticalJunctionType.HEAP_EXTRACT,
      junctionDifficulty: JunctionDifficulty.PROCEDURAL,
    })

    const temp = working[0]
    working[0] = working[i]
    working[i] = temp

    push({
      description: `Swapped the maximum into its final sorted position at index ${i}. The array is now [${working.join(', ')}].`,
      pseudocodeLine: PSEUDOCODE_LINE.EXTRACT_SWAP,
      isPredictionRequired: false,
      state: { array: working, heapSize: i },
      swappedIndices: [0, i],
      highlightIndices: Array.from({ length: n - i }, (_, idx) => i + idx),
    })

    push({
      description: `Sifting the new root down to restore the heap property within the remaining ${i} elements.`,
      pseudocodeLine: PSEUDOCODE_LINE.SIFT_DOWN,
      isPredictionRequired: false,
      state: { array: working, heapSize: i },
      highlightIndices: Array.from({ length: n - i }, (_, idx) => i + idx),
    })
    heapify(0, i)
  }

  push({
    description: 'The array is fully sorted. What invariant proves that sorting is complete?',
    pseudocodeLine: PSEUDOCODE_LINE.DONE,
    isPredictionRequired: true,
    state: { array: working, heapSize: 0 },
    highlightIndices: Array.from({ length: n }, (_, idx) => idx),
    criticalJunctionType: CriticalJunctionType.ALGORITHM_COMPLETE,
    junctionDifficulty: JunctionDifficulty.CONCEPTUAL,
  })

  push({
    description: `Heap Sort complete. The array is fully sorted: [${working.join(', ')}].`,
    pseudocodeLine: PSEUDOCODE_LINE.DONE,
    isPredictionRequired: false,
    state: { array: working, heapSize: 0 },
    highlightIndices: Array.from({ length: n }, (_, idx) => idx),
    isFinalStep: true,
  })

  return snapshots
}

File: apps/ai/prompts/heap_sort.py
HEAP_SORT_CONTEXT = """
Heap Sort works in two phases on a single array. First, it builds a
max-heap in place: every internal node (from the last non-leaf down to the
root) is "heapified" - compared against its larger child and swapped down
if it violates the heap property (a parent must be >= both children),
repeating at the new position until the property holds or a leaf is
reached. Once every internal node has been heapified bottom-up, the whole
array satisfies the max-heap property and the maximum value sits at
index 0.

Second, it repeatedly extracts the maximum: swap the root (index 0) with
the last element still in the unsorted region, shrink that region by one
(the swapped-in value is now in its final sorted position), then sift the
new root down to restore the heap property within the smaller region.
Repeating this until only one element remains leaves the array fully
sorted in place.

When giving feedback, always refer to the actual index numbers and values
currently in the array, and be clear about which phase (heapify or
extract) the student is in.

Key properties:
- Time complexity: O(n log n) in all cases - building the heap is O(n),
  and each of the n extractions costs O(log n) to sift down.
- Not stable - equal elements can be reordered by the heap operations.
- In-place: O(1) extra space, unlike Merge Sort.

Common errors to watch for:
- ORDER_OF_OPERATIONS: The student says a parent should sift down when it
  is already at least as large as both children, or the reverse. Ask them
  to compare the parent against its larger child again.
- STRUCTURAL_PROPERTY_VIOLATION: The student thinks the heap is fully
  sorted and predicts index 1 holds the second-largest value - only the
  parent/child relationship is guaranteed, not the order between siblings
  or across subtrees.
"""

CRITICAL_JUNCTION_GUIDANCE: dict[str, str] = {
    "HEAP_COMPARE": (
        "The student was asked whether a parent node should sift down "
        "past its larger child. Ground the feedback in the exact "
        "parent and child values and indices being compared."
    ),
    "HEAP_EXTRACT": (
        "The student was confirming that the current root is the "
        "maximum about to be extracted into its final sorted position. "
        "Ground the feedback in the root's exact value and how many "
        "elements remain in the unsorted region."
    ),
    "ALGORITHM_COMPLETE": (
        "The student was asked to identify the invariant that proves "
        "sorting is complete. The correct invariant is that no "
        "adjacent pair is out of order - every element has been "
        "extracted into its final sorted position. Ground the feedback "
        "in this invariant, not in any single comparison."
    ),
}

HEAP_SORT_PSEUDOCODE = """1  procedure heapSort(A: list)
2      n = length(A)
3      for i = floor(n / 2) - 1 downto 0 do
4          heapify(A, n, i)
5      end for
6      for end = n - 1 downto 1 do
7          swap(A[0], A[end])
8          heapify(A, end, 0)
9      end for
10 end procedure
11
12 procedure heapify(A, size, i)
13     largest = i; left = 2i + 1; right = 2i + 2
14     if left < size and A[left] > A[largest] then largest = left
15     if right < size and A[right] > A[largest] then largest = right
16     if largest != i then
17         swap(A[i], A[largest])
18         heapify(A, size, largest)
19     end if
20 end procedure
"""
